### ISSUE TITLE
Generate typed structures

### ISSUE THREAD

**Issue by ohadschn** [kevingosse/DynaMD#1]
`dynamic` is nice, but having a typed object would be much better when possible (type safety, inteliisense, etc). I propose a method with the following signature:
```cs
T GetTypedObject<T>(this ClrObject object, bool deep = false, bool invokeCtor = false)
```
1. Depending on `invokeCtor`, either use `Activator.CreateInstance` or `FormatterServices.GetUninitializedObject` for instantiation of `T` and its fields (possibly split to `GetTypedObjectWithCtor` so you can add a `new` constraint and do `new T()`).
2. Depending on `deep`, either perform a shallow copy of the fields, or a recursive one. Terminate at known types (primitives, strings, arrays etc.) as well as previously visited objects (keep dictionary) to prevent infinite loops.

This would work especially well for types that are public and accessible to the debugging tool, but even for internal types one could just stub out the type (mocking the fields one cares about) and enjoy type safety from then on.

**Comment by kevingosse** [comment]
Currently the library supports marshalling of self-contained types (primitives types, structs containing only primitives types, and arrays of such structs. In a nutshell, everything that does not contain a reference).
I have a prototype that supports all types (using FormatterServices.GetUninitializedObject), but it breaks on generic types because of bugs in ClrMD. It's going to take a while to identify them and fix them upstream. 
I think I'll push that prototype on a separate branch in the meantime.

**Comment by ohadschn** [comment]
Not sure how bugs in ClrMD are relevant? 

If you can create the proxy object, that means you can access the fields, which means you can populate an object you got from `FormatterServices.GetUninitializedObject` just as you can populate your proxy. Once you have the fields, it's in your hands, no?

**Comment by kevingosse** [comment]
Well, I've met two kind of problems so far:

- In some cases, the ClrType isn't resolved properly by ClrMD. I need it to know the layout of the object in memory, so that I can fetch the value of the fields. 
In some cases, I've been able to bypass that by resolving the type twice: https://github.com/kevingosse/DynaMD/blob/marshalling/src/DynaMD/DynamicProxy.cs#L321  (@chrisnas actually filled a bug about this one: https://github.com/Microsoft/clrmd/issues/108)
In another case, I've been able to bypass by resolving the ClrType myself using the information from the type we're trying to cast to: https://github.com/kevingosse/DynaMD/blob/marshalling/src/DynaMD/DynamicProxy.cs#L264

- The other, that I'm still stuck on, is that some features aren't actually implemented yet in ClrMD: https://github.com/Microsoft/clrmd/blob/ecfe9790f6ef5c78afa63db04ccbe0dda15e86bf/src/Microsoft.Diagnostics.Runtime/Desktop/types.cs#L709
Basically, depending on the way I resolve the ClrType, ClrMD will either give me an instance of DesktopHeapType (which has a working GetArrayElementAddress method) or DesktopArrayType  (in which the method isn't implemented). Unfortunately, when I resolve the ClrType manually to bypass the aforementioned issues, I end up with a DesktopArrayType instance, and so I can't access the values.

That said, the layout of an array is predictable enough, so I may end up computing the items addresses in memory by myself.

If you're interested, I've pushed the proof of concept in a branch: https://github.com/kevingosse/DynaMD/tree/marshalling
You can see the issue by running the `Can_marshal_to_class5` unit test.

**Comment by ohadschn** [comment]
Thank you for explaining, but I still don't see how these issues are relevant to whether you populate a typed object or a proxy object. Whatever fields you can get for your proxy, put in the typed object...

**Comment by kevingosse** [comment]
Ok, got you. I think I've managed to bypass all the bugs left by following your advice (gotta do more tests to be sure). Instead of directly using the type information returned by ClrMD, I just retrieve the address then re-fetch the type from that address (as would be naively done if you used proxies, where a new proxy is created every time you access a field).

The last thing missing is for fields with abstract types or interfaces.

For instance, something like:

```
class MyClass
{
    public IDisposable Field;
}
```

When trying to cast the proxy to `MyClass`, I can't find the concrete type of Field by using reflection. I need to extract the ClrType from ClrMD then convert it to System.Type. The issue is that ClrMD uses a different syntax than .NET for generic types (for instance, `System.List`1[System.Int32]` for .NET, and `System.List<System.Int32>` for ClrMD). This requires writing a parser (I don't think it can be done with a regex, as generic arguments can themselves be generic). Not a big deal, I just need to find some time for it.

**Comment by ohadschn** [comment]
I see what you mean. Strictly speaking, you would need to do this for any non-sealed type...

**Comment by ohadschn** [comment]
I guess you could just put `null` in such places until you get the parser working...

**Comment by ohadschn** [comment]
BTW I left some feedback on your workshop here: https://github.com/chrisnas/SELAConference2018/issues/1
But I think @chrisnas might not be following that repo...

**Comment by kevingosse** [comment]
Oh indeed I don't think he noticed. Thanks, I pinged him.
